### ISSUE TITLE
All Splunk logs fail

### ISSUE THREAD

**Issue by initech-asynchrony** [AlanBarber/NLog.Targets.Splunk#10]
We're trying to implement this is a very simple project, but we don't seem to be able to transmit any logs to splunk. Internal nlogs are as follows:

```
2018-04-24 15:03:09.0039 Debug Targets for Microsoft.AspNetCore.Hosting.Internal.HostedServiceExecutor by level:
2018-04-24 15:03:09.0039 Debug Trace => StdOut SplunkLog
2018-04-24 15:03:09.0051 Debug Debug => StdOut SplunkLog
2018-04-24 15:03:09.0051 Debug Info => StdOut SplunkLog
2018-04-24 15:03:09.0055 Debug Warn => StdOut SplunkLog
2018-04-24 15:03:09.0055 Debug Error => StdOut SplunkLog
2018-04-24 15:03:09.0055 Debug Fatal => StdOut SplunkLog
2018-04-24 15:03:13.1713 Error Error has been raised. Exception: System.ArgumentException: An item with the same key has already been added. Key: Host
   at System.ThrowHelper.ThrowAddingDuplicateWithKeyArgumentException(Object key)
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at NLog.Targets.Splunk.SplunkHttpEventCollector.Write(LogEventInfo logEventInfo)
   at NLog.Targets.Target.Write(AsyncLogEventInfo logEvent)
```
Our config is as follows:

```
<?xml version="1.0" encoding="utf-8" ?>
<nlog xmlns="http://www.nlog-project.org/schemas/NLog.xsd" 
      xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" 
      autoReload="true" 
      throwConfigExceptions="true"
      internalLogLevel="Debug" 
      internalLogFile="nlog-internal.log">

    <extensions>
        <add assembly="NLog.Targets.Splunk"/>
    </extensions>

    <targets>
        <target 
        xsi:type="Console"
        name="StdOut"
        encoding="ascii" 
        />

        <target
        name="SplunkLog"
        xsi:type="SplunkHttpEventCollector"
        ServerUrl="https://input-prd-p-ntx82fn9qxxx.cloud.splunk.com:8088"
        Token="XXXXXXXX-XXXX-XXXX-XXXX-9614B5702274"
        RetriesOnError="0"
        layout="${message} ${exception:format=tostring}"
        />
    </targets>

    <rules>
        <logger name="*" minlevel="Trace" writeTo="StdOut" />
        <logger name="*" minlevel="Trace" writeTo="SplunkLog" />
    </rules>
</nlog>
```

The plugin appears to load, but any sort of logging produces the same error listed above, regardless of where the call originated from within the program.

We are using AspNetCore

**Comment by snakefoot** [comment]
Think this is the problem line:

https://github.com/AlanBarber/NLog.Targets.Splunk/blob/8e8e181c87ae6bc1acc0e78cda81b6ed6a8ade53/src/NLog.Targets.Splunk/SplunkHttpEventCollector.cs#L76

Where it performs a conditional add, which conflicts with the "Host"-key already added, and exception is thrown.

Guess the really nice fix, would be that SplunkHttpEventCollector  would inherit from `TargetWithContext` (Instead of `TargetWithLayout`). And then just used `TargetWithContext.GetAllProperties` (Instead of adding "Host" and "Source" manually then they could just be added using `TargetWithContext.ContextProperties`)

**Comment by initech-asynchrony** [comment]
I was just about to post that one. That definitely seems to be the issue.

**Comment by AlanBarber** [comment]
I'm working on taking into account the changes @snakefoot suggest to use the TargetWithContext.

@initech-asynchrony can you provide a small code sample how you are using the .net core DI to create your logger and a sample log message you are performing?
I'm going to create a new aspnet core test app in the project to validate things will works correctly.

**Comment by AlanBarber** [comment]
One thing reviewing these changes is that I think the inclusion of the host and source to the properties was a mistake. This initial code used was part on a client project and they wanted to include it in the properties list so I had carried it forward when I open sourced the code. However as I've seen here it's run into conflicts if there are existing log events using the "Host" or "Source" keys in their own logs.

As seen in the logging of Microsoft.AspNetCore.Hosting.Internal.WebHost for example that logs the following log entry
```
{  
  "Level":"Info",
  "MessageTemplate":"Request starting HTTP/1.1 GET http://localhost:64499/  ",
  "RenderedMessage":"Request starting HTTP/1.1 GET http://localhost:64499/  ",
  "Properties":{  
    "Protocol":"HTTP/1.1",
    "Method":"GET",
    "ContentType":null,
    "ContentLength":null,
    "Scheme":"http",
    "Host":"localhost:64499",
    "PathBase":"",
    "Path":"/",
    "QueryString":"",
    "EventId_Id":1,
    "EventId_Name":null,
    "EventId":{  
      "Id":1,
      "Name":null
    }
  }
}
```

So, I'm ripping that out and moving to using NLog's built in tooling to getting properties.

**Comment by bruceharrison1984** [comment]
I can post a detailed example if you like, but we are using NLog exactly as stated in the new documentation:

https://github.com/NLog/NLog.Web/wiki/Getting-started-with-ASP.NET-Core-2#4-update-programcs

**Comment by AlanBarber** [comment]
@bruceharrison1984 Thank you I setup a test project following those steps and was able to reproduce the error. I have some other code changes and testing to do but I will attempt to push out a 2.0.1 pre-release nuget package shortly with fixes to the issue.

**Comment by bruceharrison1984** [comment]
Glad i could help. Look forward to using this in the future, thanks for the lightning fast turn around!

**Comment by AlanBarber** [comment]
@snakefoot I merged in PR #11 but upon running my local tests it looks like your changes have caused some small bug with the serialization changes that as a result the splunk server is throwing back a Bad Request.

**Comment by snakefoot** [comment]
@AlanBarber Very strange. The ByteArrayContent should be identical to the old StringContent. Any output from the NLog InternalLogger: https://github.com/NLog/NLog/wiki/Internal-Logging

**Comment by snakefoot** [comment]
Have created #13 that ensures the use of `JsonConvert.DefaultSettings` as initial configuration settings.

**Comment by snakefoot** [comment]
@AlanBarber Found the difference. The use of StreamWriter made it automatically add UTF BOM. Have now removed the BOM

**Comment by AlanBarber** [comment]
I've pushed out a pre-release today for test!

Please let me know if you find any issues.

https://www.nuget.org/packages/NLog.Targets.Splunk/2.1.0-rc01

**Comment by AlanBarber** [comment]
@initech-asynchrony @bruceharrison1984 Have you had a chance to test the rc01 release yet? 

If it appears to be working fine I would like to go ahead and do an official 2.1.0 release this weekend.

**Comment by bruceharrison1984** [comment]
We should be able to give this another shot this week, and I will get you
some feedback once we run through it.

**Comment by snakefoot** [comment]
@initech-asynchrony Instead of jumping into the deep end of the pool, then you should start with something simple:

- Start making the old version work in your environment using a simple net45-console-application (no asp-net-core stuff, that causes conflicts in dictionary-add)
- Then upgrade to the new updated version, and see it continues to work.
- Then move the working solution into the asp-net-core world.

**Comment by initech-asynchrony** [comment]
We tried running the included aspnetcore example project and still are not seeing any logs appear in Splunk. We started up Charles and did some HTTP traffic watching, and we never actually see the POSTs from NLog.Target.Splunk being sent out. We can see the entries in the internal NLog logs. Testing directly with Postman, we can see the outbound requests, and also see the results appear in Splunk. We're still investigating.

**Comment by snakefoot** [comment]
@initech-asynchrony No error-reply is detected by the Splunk-HttpClient, so I guess one have to compare two identical logevents with fiddler, and look for the difference (between NLog and Postman).

**Comment by bruceharrison1984** [comment]
Our problems seem to be the result of a few issues:

- We are using Self-Managed Cloud Splunk
  - This version of Splunk doesn't used signed-certificates
  - HttpClient will blow up internally unless SSL verification is disabled
  - https://stackoverflow.com/a/46626858
- HttpClient has some issues on OSX
  - The **ServerCertificateCustomValidationCallback** on the HttpClientHandler is never invoked
    - This means the normal way(see above) of disabling SSL verification doesn't work 
  - https://github.com/dotnet/corefx/issues/9728
  - This can be bypassed by setting the ServerCertificateCustomValidationCallback = HttpClientHandler. DangerousAcceptAnyServerCertificateValidator...
    - that property is only available in netcore applications, not netstandard apps.
  - https://msdn.microsoft.com/en-us/library/system.net.http.httpclienthandler.dangerousacceptanyservercertificatevalidator(v=vs.110).aspx

The catch block [here](https://github.com/AlanBarber/NLog.Targets.Splunk/blob/04d687ee30c79576a92bacd662daac88748b80f8/src/NLog.Targets.Splunk/Splunk.Logging.Common/HttpEventCollectorSender.cs#L425) will actually blow up again if e.Response is null due to an internal error. This is what was causing all of our errors to be swallowed, and not show up in the NLog internal logs. We got around this by setting the [default HttpStatusCode to InternalServerError](https://github.com/AlanBarber/NLog.Targets.Splunk/blob/c47de408ee49c31c66f0bd6c6620f98e0eda2c4d/src/NLog.Targets.Splunk/Splunk.Logging.Common/HttpEventCollectorSender.cs#L376), which seems more correct anyway. Once that change was made, we could see all of the internal HttpClient errors that were causing our problems.

We still haven't found a solution, since logs still don't appear to be transmitting, but figured I'd give you some info.

**Comment by AlanBarber** [comment]
So we're down to SSL issues... as you've seen in your own research dealing with self signed cert validation across different versions of the .NET framework is pretty much a mess right now.  I'll see what code changes we can make to hopefully more gracefully handle this.

As an aside, you can buy a cert and if you create a splunk support request you can ask them to install the cert from what I've read.

**Comment by snakefoot** [comment]
Created a random shot in the dark for ignoring ssl certificate errors: #15

**Comment by bruceharrison1984** [comment]
I may have some time tomorrow to give this a run and report back. Not sure if it'll work simply because the ServerCertificateCustomValidationCallback never appears to be invoked, but I'm willing to give it another shot.

I'd also consider leaving in the SSL verification override. Anyone else using Splunk Cloud Self-Service or a locally hosted Splunk could very well be using self-signed certs and have no legitimate reason to purchase a real certificate. It seems like a valid use case.

**Comment by snakefoot** [comment]
Maybe try using NetCore 2.1-rc:

https://github.com/dotnet/corefx/issues/17226

**Comment by bruceharrison1984** [comment]
I believe that hooking the DangerousAcceptAnyServerCertificateValidator would work, however that is only available to netcore apps, and this plugin is written as netstandard, so the property isn't available.

**Comment by bruceharrison1984** [comment]
@snakefoot  Update on your PR, it doesn't work in OSX (no surprise there), but it does give a much better error message in the Nlog internal log:

    Exception: Splunk.Logging.HttpEventCollectorException: Failed to send request: InternalServerError ---> System.PlatformNotSupportedException: The handler does not support custom handling of certificates with this combination of libcurl (7.54.0) and its SSL backend ("SecureTransport").

So in that respect, I think it is a great addition because this would have surfaced our problem immediately.

At compile time we use the MS provided container **microsoft/dotnet:2.1-sdk**. We repackaged and uploaded your PR to our internal nuget server and pull it down from there. All that works great.

At run-time we use the MS provided **microsoft/aspnetcore:2**. Running your PR code actually transmits logs, so things are now working correctly. Running with internal logs generates a ton of spam due to the logging mechanism, but it gives good proof that things are working.

So in short, it works fine on the debian distro, but is still busted on OSX due to crappy HttpClient implementation. Thanks!

**Comment by AlanBarber** [comment]
Pushed out RC03 with latest IgnoreSslErrors and updated NLog version

https://www.nuget.org/packages/NLog.Targets.Splunk/2.1.0-rc03

**Comment by bruceharrison1984** [comment]
Were running RC03 now and all appears to be working fine

**Comment by snakefoot** [comment]
Have one minor fix, and then I think ver. 2.1.0 is ready. See #16

**Comment by AlanBarber** [comment]
closing things out and will be doing a official 2.1.0 release today.

thanks everyone !
